aspects/aspects.go

```diff
@@ -0,0 +1,343 @@
+		Name:    name,
+		dataBag: dataBag,
+		schema:  schema,
+		aspects: make(map[string]*Aspect, len(aspects)),
+	}
+
+	for name, v := range aspects {
+		aspectPatterns, ok := v.([]map[string]string)
+		if !ok {
+			return nil, errors.New("cannot create aspect: access patterns should be a list of maps")
+		} else if len(aspectPatterns) == 0 {
+			return nil, errors.New("cannot create aspect without access patterns")
+		}
+
+		aspect := &Aspect{
+			Name:           name,
+			accessPatterns: make([]*accessPattern, 0, len(aspectPatterns)),
+			directory:      aspectDir,
+		}
+
+		for _, aspectPattern := range aspectPatterns {
+			name, ok := aspectPattern["name"]
+			if !ok || name == "" {
+				return nil, errors.New(`cannot create aspect pattern without a "name" field`)
+			}
+
+			// TODO: validate that a path isn't a subset of another (possibly somewhere else).
+			// Otherwise,  we can write a user value in a subkey of a path (that should be a map).
+			path, ok := aspectPattern["path"]
+			if !ok || path == "" {
+				return nil, errors.New(`cannot create aspect pattern without a "path" field`)
+			}
+
+			access, err := newAccessType(aspectPattern["access"])
+			if err != nil {
+				return nil, fmt.Errorf("cannot create aspect pattern: %w", err)
+			}
+
+			aspect.accessPatterns = append(aspect.accessPatterns, &accessPattern{
+				name:   name,
+				path:   path,
+				access: access,
+			})
+		}
+
+		aspectDir.aspects[name] = aspect
+	}
+
+	return &aspectDir, nil
+}
+
+// Aspect returns an aspect from the aspect directory.
+func (d *Directory) Aspect(aspect string) *Aspect {
+	return d.aspects[aspect]
+}
+
+// Aspect is a group of access patterns under a directory.
+type Aspect struct {
+	Name           string
+	accessPatterns []*accessPattern
+	directory      Directory
+}
+
+// Set sets the named aspect to a specified value.
+func (a *Aspect) Set(name string, value interface{}) error {
+	for _, p := range a.accessPatterns {
+		if p.name != name {
+			continue
+		}
+
+		if !p.isWriteable() {
+			return fmt.Errorf("cannot set %q: path is not writeable", name)
+		}
+
+		if err := a.directory.dataBag.Set(p.path, value); err != nil {
+			return err
+		}
+
+		data, err := a.directory.dataBag.Data()
+		if err != nil {
+			return err
+		}
+
+		return a.directory.schema.Validate(data)
+
+	}
+
+	return &NotFoundError{fmt.Sprintf("cannot set %q: name not found", name)}
+}
+
+// Get returns the aspect value identified by the name. If either the named aspect
+// or the corresponding value can't be found, a NotFoundError is returned.
+func (a *Aspect) Get(name string, value interface{}) error {
+	for _, p := range a.accessPatterns {
+		if p.name != name {
+			continue
+		}
+
+		if !p.isReadable() {
+			return fmt.Errorf("cannot get %q: path is not readable", name)
+		}
+
+		if err := a.directory.dataBag.Get(p.path, value); err != nil {
+			if errors.Is(err, &NotFoundError{}) {
+				return &NotFoundError{fmt.Sprintf("cannot get %q: %v", name, err)}
```

Comment:
Another silly question, why is this wrapped again, why can't the underlying layer give us enough context here already?

Reply:
name vs path I think. We actually need to think a bit how much we want to expose or not expose paths in errors. I need to re-review this for that

Reply:
I still would like to do a review of the latest changes, I will do it later today or first thing in the morning tomorrow